src/loan.rs

```diff
@@ -417,28 +469,46 @@ impl Borrower1 {
 
         // fulfill collateral input covenant script
         {
-            let sighash = SigHashCache::new(&tx).segwitv0_sighash(
-                1,
-                &self.collateral_script.clone(),
-                self.repayment_collateral_input.original_txout.value,
-                SigHashType::All,
-            );
+            let descriptor = self.collateral_descriptor.clone();
+            let descriptor_cov = descriptor.as_cov()?;
 
-            let sig = SECP256K1.sign(
-                &elements::secp256k1_zkp::Message::from(sighash),
-                &self.keypair.0,
-            );
+            let collateral_value = self.repayment_collateral_input.original_txout.value;
+            let cov_script = descriptor_cov.cov_script_code();
 
-            let script_witness = RepaymentWitnessStack::new(
-                sig,
-                self.keypair.1,
-                self.repayment_collateral_input.original_txout.value,
-                &tx,
-                self.collateral_script.clone(),
-            )
-            .unwrap()
-            .serialize()
-            .unwrap();
+            let cov_sat =
+                CovSatisfier::new_segwitv0(&tx, 1, collateral_value, &cov_script, SigHashType::All);
+
+            let cov_pk_sat = {
+                let mut hash_map = HashMap::new();
+                let sighash = cov_sat.segwit_sighash()?;
+                let sighash = elements::secp256k1_zkp::Message::from(sighash);
+
+                let sig = SECP256K1.sign(&sighash, &COVENANT_SK);
+                hash_map.insert(*descriptor_cov.pk(), (sig, SigHashType::All));
+
+                hash_map
+            };
+
+            let ident_pk_sat = {
+                let mut hash_map = HashMap::new();
+
+                let script = descriptor.explicit_script();
+                let sighash = SigHashCache::new(&tx).segwitv0_sighash(
+                    1,
+                    &script,
+                    collateral_value,
+                    SigHashType::All,
+                );
+                let sighash = elements::secp256k1_zkp::Message::from(sighash);
+
+                let sig = SECP256K1.sign(&sighash, &self.keypair.0);
+                hash_map.insert(self.keypair.1, (sig, SigHashType::All));
+
+                hash_map
+            };
+
+            let (script_witness, _) =
+                descriptor_cov.get_satisfaction((cov_sat, cov_pk_sat, ident_pk_sat))?;
```

Comment:
Can you not call `satisfy` instead and have it populate the witness stack for you?

Reply:
Yep, that's a better API. But it turns out that the `CovSatisfier` takes an immutable reference to `Transaction` and `satisfy` then takes a mutable reference to it. I know how to get around it by cloning the transaction an passing that one to the `CovSatisfier`, but it seems wasteful. Perhaps there exists an alternative that I'm missing.